File: math/range_product_query_of_powers.rb
# @param {Integer} n
# @param {Integer[][]} queries
# @return {Integer[]}
def product_queries(n, queries)
    binary = []
    rep = 1
    mod = 10**9 + 7

    while n != 0
        binary << rep if n % 2 == 1
        n /= 2
        rep *= 2
    end

    queries.map do |l, r|
        prod = 1
        (l..r).each { |i| prod = (prod * binary[i]) % mod }
        prod
    end
end

File: math/range_product_query_of_powers.py
.class Solution:
    def productQueries(self, n: int, queries: List[List[int]]) -> List[int]:
        binary = []
        mod = 10 ** 9 + 7
        rep = 1
        while n != 0:
            if n % 2 == 1:
                binary.append(rep)
            n = n // 2
            rep *= 2

        ans = []
        prods = []

        prod = 1
        for rep in binary:
            prod *= rep
            prods.append(prod)

        for l, r in queries:
            ans.append((prods[r + 1] // prods[l]) % mod)
        return ans